jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,584 @@
+   * if needed (moving the tail to bottom_levels if any)
+   * - finally, updates qty at the relative(px)
+   */
+  bool add_order(price4_t px, int qty) {
+    validate_add_order_params(qty, px);
+    // get px price levels
+    auto tk_px = price_levels(price4_t(0), px);
+    // check if tk_px is worse than the first price of the top_levels_
+    if (side<compare_t>::better_level(tk_begin_top_, tk_px)) {
+      // emplace the price at the bottom_levels, and return (false)
+      auto emp_tup = bottom_levels_.emplace(tk_px, 0);
+      emp_tup.first->second += qty;
+      return false;
+    }
+    // check if tk_px is equal or better than the current inside
+    if (not side<compare_t>::better_level(tk_inside_, tk_px)) {
+      // ... check if limit redefine is needed
+      if (not side<compare_t>::better_level(tk_end_top_, tk_px)) {
+        // get the new limits based on the new inside tk_px
+        auto limits = side<compare_t>::limit_top_prices(tk_px, max_size_ / 2);
+        // move the tail [px_begin_top_, new px_begin_top_) to bottom_levels_
+        move_top_to_bottom(std::get<0>(limits));
+        // ... redefine the limits
+        tk_begin_top_ = std::get<0>(limits);
+        tk_end_top_ = std::get<1>(limits);
+      }
+      // if the tk_px inside changed, updated
+      if (tk_inside_ != tk_px) {
+        tk_inside_ = tk_px;
+      }
+      // update top_levels_
+      auto rel_px =
+          side<compare_t>::level_to_relative(tk_begin_top_, tk_inside_);
+      top_levels_.at(rel_px) += qty;
+      return true; // the inside changed
+    }
+    // is a top_levels change different than the inside
+    // udates top_levels_ with new qty
+    auto rel_px = side<compare_t>::level_to_relative(tk_begin_top_, tk_px);
+    top_levels_.at(rel_px) += qty;
+    return false;
+  }
+
+  /**
+   * Reduce the quantity for a given price.
+   *
+   * @param px the price of the order that was reduced
+   * @param qty the quantity reduced in the order
+   * @returns true if the inside changed
+   *
+   * @throw feed_error qty <= 0
+   * @throw feed_error px should be in bottom_levels_ but is empty
+   * @throw feed_error trying to reduce a non-existing bottom_levels price
+   * @throw feed_error trying to reduce a non-existing top_levels price
+   * @throw feed_error trying to reduce a price better than px_inside_
+   *
+   * Checks and handles if px is a bottom_level_ price
+   * Checks if top_levels_ is empty after reducing
+   * If so, limits have to be redefined, and tail moved into top_levels
+   */
+  bool reduce_order(price4_t px, int qty) {
+    validate_add_order_params(qty);
+    // get px price level
+    auto tk_px = price_levels(price4_t(0), px);
+    // check and handles if it is a bottom_level_ price
+    if (side<compare_t>::better_level(tk_begin_top_, tk_px)) {
+      auto price_it = bottom_levels_.find(tk_px);
+      if (price_it == bottom_levels_.end()) {
```

Comment:
ACK. I need more info here.

Option 1: create a handle_exception function.-
This code 
`
        std::ostringstream os;
        os << "array_based_order_book::reduce_order."
           << " Trying to reduce a non-existing bottom_levels_ price."
           << " tk_begin_top_=" << tk_begin_top_ << " px=" << px
           << " qty=" << qty;
        throw jb::feed_error(os.str());
`
changed by:
`
  handle_exception(str, param1, param2, ... /*yes, eliptic to use it in other part of code */
`

OR

Option 2: Create reduce_bottom_levels(px, qty)
Mode all the code from 233 to 252 right before:: return false;
If this is the case, I need help to understand how the template class is going to be defined if this code is into a cpp file.

Please clarify.

Waiting for comments!!!